### PR TITLE
Helper decrypted

### LINKED ISSUE #35: Image to Google Books
Need to design a system to get text from an image to hand to Google Books to make a dynamic search. This will take some time.

### PR DESCRIPTION
**Pull request by David1000734**
Added a base64 string descriptor used for getting an image. If assuming a valid input for string64, the function was tested and successfully created an image from the string. Error checking was adding to a certain margin for the function.

## Description
Added a helper function that will handle parsing the base64 string into an image. Notably, this function will reconstruct an actual .png file and should be properly dealt with after it is used.

## Type of Change

- [ ] Bugfix
- [x] New Feature
- [ ] Documentation Update
- [ ] Other (please describe):

## Checklist

- [ ] I have read the contributing guidelines.
- [x] I have followed the code style of this project.
- [x] I have added tests to cover my changes.
- [ ] I have updated the documentation (if applicable).
- [x] My changes do not generate new warnings.

## Related Issue
Addresses #19, tests are included in this pull request.
Closes #35, solution for turning the string into images for the server.

### CODE REVIEW

**Review by LilAttacker — APPROVED** [review]
Looks great. No issues. Tests are included and this is covered. It looks good as a beginning.

This will eventually need more added to it (like the ability to read in a parameter to decrypt), calls made to image recognition from this, etc. But as a base and starting point, it looks good.

**Comment by David1000734** [comment]
> Looks great. No issues. Tests are included and this is covered. It looks good as a beginning.
> 
> This will eventually need more added to it (like the ability to read in a parameter to decrypt), calls made to image recognition from this, etc. But as a base and starting point, it looks good.

Right now the function takes three parameters
1. base64 string <Image Data>
2. File Name
3. Path

As it stands, it should be able to handle the parameter passing for decrypting other images. I envisioned this function to be called before the image recognition or any other models so that an image is created at a specified path. This path can then be provided to the image recognition to find it and do stuff with it, then finally, the image created will be deleted by the programmer at the end of the function. 

If you think there is a better way to go about doing this, I can make the changes since this would be a function that you guys are using.

File: server/tests/test_helper.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import tabby_server.services.helper as helper
+from tests.variable import test_base64
+import os
+import unittest
+
+
+class Helper (unittest.TestCase):
```
Comment (ederic-oytas) [inline server/tests/test_helper.py:7]: We use `pytest`, so test functions do not need to be in a class. Pytest runs all functions which start with `test_`.

**Review by ederic-oytas — CHANGES_REQUESTED** [review]
Hi David, good job on finding a way to convert the image. However, I don't think writing it to a file is the most efficient solution; it needs to converted directly into an image matrix. There is a function called `cv2.imdecode` from OpenCV which you can use. Here's a ChatGPT conversation about it: https://chatgpt.com/share/67319fa3-3f2c-8009-b3b4-ec476c464dac

Here are also some other changes I want to see:
- I don't think this function belongs in `services/`. The image will be decoded when it's received as a request, so it should belong in `api/`.
- I think there is a better name than `helper.py`.
- Remove the type checks (see my comment in helper.py)
- We use pytest, not unittest (see my comment in test_helper.py)

**Comment by ederic-oytas** [comment]
By the way, remember the encoding. I believe the Base64 string, when decoded, will convert to bytes which is for a JPG file.

**Comment by David1000734** [comment]
> Hi David, good job on finding a way to convert the image. However, I don't think writing it to a file is the most efficient solution; it needs to converted directly into an image matrix. There is a function called `cv2.imdecode` from OpenCV which you can use. Here's a ChatGPT conversation about it: https://chatgpt.com/share/67319fa3-3f2c-8009-b3b4-ec476c464dac
> 
> Here are also some other changes I want to see:
> 
> * I don't think this function belongs in `services/`. The image will be decoded when it's received as a request, so it should belong in `api/`.
> * I think there is a better name than `helper.py`.
> * Remove the type checks (see my comment in helper.py)
> * We use pytest, not unittest (see my comment in test_helper.py)

I can move the function to api.

The reason I named it helper was because I assumed we can place other helper function into this py file instead of making multiple which means we have to import multiple functions if we need them. Regardless, I can name it something else; I just don't know what to call it, any suggestions?

I'll look into having the function provide an image matrix instead of just creating the png then.

**Comment by LilAttacker** [comment]
> I don't think writing it to a file is the most efficient solution; it needs to converted directly into an image matrix.

Do we actually need an image matrix though? For YOLO, it scans the full image. It's input is the image.

Is this for easyOCR? Because YOLO I think only works with the full image.

**Review by LilAttacker — CHANGES_REQUESTED** [review]
Your tests currently don't work.

Please fix them before we merge.

**Comment by ederic-oytas** [comment]
To fix your tests, I believe you have to add the `pillow` dependency with `poetry add`.

By the way, please use `poetry add pillow` instead of `pip install pillow`.
- `poetry add` installs it and adds it to the project dependencies, so it can be later installed with `poetry install`.
- `pip install` only installs it to the current environment, and it isn't tracked.

**Comment by LilAttacker** [comment]
While Ederic is likely right on the issue (and is absolutely correct in regards to `poetry add`), the actual issue raised was in the linter, Flake8, and it raised:
`tabby_server/api/base64_to_image.py:35:11: F821 undefined name 'Image'`
This should be pretty easy to fix.

File: server/tabby_server/api/base64_to_image.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    :param string64: Base 64 string. Function will not check to see if it is
+        actually a base 64 string
+    """
+    # First decode the base64 into byte data
+    imgdata = base64.decodebytes(bytes(string64, 'utf-8'))
+
+    # Process the byte data into BGR (Blue, Gree, Red) where
+    # blue occupies the most significant bit and red occupies the LSB
+    img = Image.open(io.BytesIO(imgdata))
```
Comment (LilAttacker) [inline server/tabby_server/api/base64_to_image.py:35]: Specifically this line failed the linter

File: server/tabby_server/api/base64_to_image.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import base64
+import cv2
+import io
+import numpy as np
+
+
+""" Example Useage
+    import tabby_server.api.base64_to_image as base64_to_image
+    from PIL import Image
```
Comment (LilAttacker) [inline server/tabby_server/api/base64_to_image.py:9]: You can actually see this line should solve the issue if it was not commented. Just be sure to add pillow to the poetry list.

**Review by LilAttacker — APPROVED** [review]
Good job fixing that! Approved.

**Review by ederic-oytas — APPROVED** [review]
I think this looks good. I'll just resolve some of the merging conflicts real quick then merge.

[X] Feature complete
[X] New test coverage, all tests pass
[X] Follows unified style
[X] Has documentation

**Pull request merged**